File: Constrain_Device/python-components/src/main/python/programmingtheiot/cda/app/DeviceDataManager.py
#####
# 
# This class is part of the Programming the Internet of Things project.
# 
# It is provided as a simple shell to guide the student and assist with
# implementation for the Programming the Internet of Things exercises,
# and designed to be modified by the student as needed.
#

import logging

from programmingtheiot.cda.connection.CoapClientConnector import CoapClientConnector
from programmingtheiot.cda.connection.MqttClientConnector import MqttClientConnector

from programmingtheiot.cda.system.ActuatorAdapterManager import ActuatorAdapterManager
from programmingtheiot.cda.system.SensorAdapterManager import SensorAdapterManager
from programmingtheiot.cda.system.SystemPerformanceManager import SystemPerformanceManager

import programmingtheiot.common.ConfigConst as ConfigConst
from programmingtheiot.common.ConfigUtil import ConfigUtil

from programmingtheiot.common.IDataMessageListener import IDataMessageListener
from programmingtheiot.common.ISystemPerformanceDataListener import ISystemPerformanceDataListener
from programmingtheiot.common.ITelemetryDataListener import ITelemetryDataListener
from programmingtheiot.common.ResourceNameEnum import ResourceNameEnum

from programmingtheiot.data.DataUtil import DataUtil
from programmingtheiot.data.ActuatorData import ActuatorData
from programmingtheiot.data.SensorData import SensorData
from programmingtheiot.data.SystemPerformanceData import SystemPerformanceData

from programmingtheiot.cda.connection.CoapClientConnector import CoapClientConnector

class DeviceDataManager(IDataMessageListener):
	"""
	Shell representation of class for student implementation.
	
	"""
	
	def __init__(self):
		# pass
		self.configUtil = ConfigUtil()

		self.enableSystemPerf = self.configUtil.getBoolean(section = ConfigConst.CONSTRAINED_DEVICE, \
													 		key = ConfigConst.ENABLE_SYSTEM_PERF_KEY)
		
		self.enableSensing = self.configUtil.getBoolean(section = ConfigConst.CONSTRAINED_DEVICE, \
												  			key = ConfigConst.ENABLE_SENSING_KEY)
		
		self.enableActuation = self.configUtil.getBoolean(section = ConfigConst.CONSTRAINED_DEVICE,\
															key = ConfigConst.ENABLE_ACTUATION_KEY)

		self.enableMqttClient = self.configUtil.getBoolean(section = ConfigConst.CONSTRAINED_DEVICE, \
													 		key = ConfigConst.ENABLE_MQTT_CLIENT_KEY) 
		
		self.enableCoapClient = self.configUtil.getBoolean(section = ConfigConst.CONSTRAINED_DEVICE, \
													 		key = ConfigConst.ENABLE_COAP_CLIENT_KEY)
		
		
		# Declaring all sensing and actuation managers
		self.sysPerfMgr = None
		self.sensorAdapterMgr = None
		self.actuatorAdapterMgr = None

		# Declaring connection variables
		self.mqttClient = None
		self.coapClient = None
		self.coapServer = None

		# Enabling system performance manager
		if self.enableSystemPerf:
			self.sysPerfMgr = SystemPerformanceManager()
			self.sysPerfMgr.setDataMessageListener(self)
			logging.info("Local System Performance Tracking Enabled")

		# Enabling Sensing 
		if self.enableSensing:
			self.sensorAdapterMgr = SensorAdapterManager()
			self.sensorAdapterMgr.setDataMessageListener(self)
			logging.info("Local sensor tracking enabled")

		# Enabling Actuation
		if self.enableActuation:
			self.actuatorAdapterMgr = ActuatorAdapterManager()
			self.actuatorAdapterMgr.setDataMessageListener(self)
			logging.info("Local Actuation capabilities enabled")

		# Enabling Mqtt Connection
		if self.enableMqttClient:
			self.mqttClient = MqttClientConnector()
			self.mqttClient.setDataMessageListener(self)

		# Enabling CoAP Client Connection
		if self.enableCoapClient:
			self.coapClient = CoapClientConnector(dataMsgListener = self)

		"""Handeling Temperature change"""
		self.handelTempChangeOnDevice = self.configUtil.getBoolean(ConfigConst.CONSTRAINED_DEVICE, \
															 	ConfigConst.HANDLE_TEMP_CHANGE_ON_DEVICE_KEY)
		
		self.triggerHvacTempFloor = self.configUtil.getFloat(ConfigConst.CONSTRAINED_DEVICE, \
													   		ConfigConst.TRIGGER_HVAC_TEMP_FLOOR_KEY)
		
		self.triggerHvacTempCeiling = self.configUtil.getFloat(ConfigConst.CONSTRAINED_DEVICE, \
														 	ConfigConst.TRIGGER_HVAC_TEMP_CEILING_KEY)
		


	def getLatestActuatorDataResponseFromCache(self, name: str = None) -> ActuatorData:
		"""
		Retrieves the named actuator data (response) item from the internal data cache.
		
		@param name
		@return ActuatorData
		"""
		pass
		
	def getLatestSensorDataFromCache(self, name: str = None) -> SensorData:
		"""
		Retrieves the named sensor data item from the internal data cache.
		
		@param name
		@return SensorData
		"""
		pass
	
	def getLatestSystemPerformanceDataFromCache(self, name: str = None) -> SystemPerformanceData:
		"""
		Retrieves the named system performance data from the internal data cache.
		
		@param name
		@return SystemPerformanceData
		"""
		pass
	

	def handleActuatorCommandMessage(self, data: ActuatorData = None) -> ActuatorData:
		"""
		This callback method will be invoked by the connection that's handling
		an incoming ActuatorData command message.
		
		@param data The incoming ActuatorData command message.
		@return boolean
		"""
		logging.info("Actuator data: "+ str(data))

		if data:
			logging.info("Processing actuator command message\n")
			return self.actuatorAdapterMgr.sendActuatorCommand(data)
		else:
			logging.warning("Incoming actuator command is invalid (null). Ignoring.")
			return None
	

	def handleActuatorCommandResponse(self, data: ActuatorData = None) -> bool:
		"""
		This callback method will be invoked by the actuator manager that just
		processed an ActuatorData command, which creates a new ActuatorData
		instance and sets it as a response before calling this method.
		
		@param data The incoming ActuatorData response message.
		@return boolean
		"""
		# pass
		if data:
			logging.debug("Incoming actuator response received (from actuator manager): "+ str(data))

			# storing data in cache
			# self.actuatorResponseCache[data.getName()] = data

			# Converting Actuator data to JSON and get the message resource
			actuatorMsg = DataUtil().actuatorDataToJson(data)
			resourceName = ResourceNameEnum.CDA_ACTUATOR_CMD_RESOURCE

			# Delegate to the transmit function any potential upstream comm's
			self._handleUpstreamTransmission(resource = resourceName, msg = actuatorMsg)

			return True
		else:
			logging.warning("Incoming actuator response is invalid (null). Ignoring..")
			return False
	

	def handleIncomingMessage(self, resourceEnum: ResourceNameEnum, msg: str) -> bool:
		"""
		This callback method is generic and designed to handle any incoming string-based
		message, which will likely be JSON-formatted and need to be converted to the appropriate
		data type. You may not need to use this callback at all.
		
		@param data The incoming JSON message.
		@return boolean
		"""
		pass
	
	def handleSensorMessage(self, data: SensorData = None) -> bool:
		"""
		This callback method will be invoked by the sensor manager that just processed
		a new sensor reading, which creates a new SensorData instance that will be
		passed to this method.
		
		@param data The incoming SensorData message.
		@return boolean
		"""
		# pass
		if data:
			logging.info("Incomming Sensor data received (from sensor manager): " + str(data))
			self._handleSensorDataAnalysis(data)

			#Converting SensorData into JSON
			jsonData = DataUtil().sensorDataToJson(data = data)

			# Passing(Upstream to GDA) Transmitting the json data to GDA via connection established
			self._handleUpstreamTransmission(resourceName = ResourceNameEnum.CDA_SENSOR_MSG_RESOURCE, msg = jsonData)

			return True
		else:
			logging.warning("Incoming sensor data is invalid (null). Ignoring..")
			return false

	
	def handleSystemPerformanceMessage(self, data: SystemPerformanceData = None) -> bool:
		"""
		This callback method will be invoked by the system performance manager that just
		processed a new sensor reading, which creates a new SystemPerformanceData instance
		that will be passed to this method.
		
		@param data The incoming SystemPerformanceData message.
		@return boolean
		"""
		# pass
		if data:
			logging.debug("Incoming system performance message received (from sys perf manager): " + str(data))
			
			jsonData = DataUtil().systemPerformanceDataToJson(data = data)
			self._handleUpstreamTransmission(resourceName = ResourceNameEnum.CDA_SYSTEM_PERF_MSG_RESOURCE, msg = jsonData)
			return True
		else:
			logging.warning("Incoming system performance data is invalid (null). Ignoring..")
			return False
		

	
	def setSystemPerformanceDataListener(self, listener: ISystemPerformanceDataListener = None):
		pass
			
	def setTelemetryDataListener(self, name: str = None, listener: ITelemetryDataListener = None):
		pass
			
	def startManager(self):
		# pass
		logging.info("*****Starting Device Data Manager******")

		# Starting System Performance Manager
		if self.sysPerfMgr:
			self.sysPerfMgr.startManager()

		# Starting Sensor Adaptor Manager
		if self.sensorAdapterMgr:
			self.sensorAdapterMgr.startManager()

		# Starting and Connecting Mqtt Client
		if self.mqttClient:
			self.mqttClient.connectClient()
			self.mqttClient.subscribeToTopic(ResourceNameEnum.CDA_ACTUATOR_CMD_RESOURCE, callback = None, qos = ConfigConst.DEFAULT_QOS)


		logging.info("Started Device Data Manager Successfully")
		
	def stopManager(self):
		# pass
		logging.info("*****Stopping Device Data Manager*****")

		# Stopping System Performance Manager
		if self.sysPerfMgr:
			self.sysPerfMgr.stopManager()
		
		# Stopping Sensor Adaptor Manager
		if self.sensorAdapterMgr:	
			self.sensorAdapterMgr.stopManager()

		# Disconnecting and Stopping Mqtt Client
		if self.mqttClient:
			self.mqttClient.unsubscribeFromTopic(ResourceNameEnum.CDA_ACTUATOR_CMD_RESOURCE)
			self.mqttClient.disconnectClient()

		logging.info("Stopped Device Data Manager Successfully")
		
	def _handleIncomingDataAnalysis(self, msg: str):
		"""
		Call this from handleIncomeMessage() to determine if there's
		any action to take on the message. Steps to take:
		1) Validate msg: Most will be ActuatorData, but you may pass other info as well.
		2) Convert msg: Use DataUtil to convert if appropriate.
		3) Act on msg: Determine what - if any - action is required, and execute.
		"""
		pass
		
	def _handleSensorDataAnalysis(self, data: SensorData = None):
		"""
		Call this from handleSensorMessage() to determine if there's
		any action to take on the message. Steps to take:
		1) Check config: Is there a rule or flag that requires immediate processing of data?
		2) Act on data: If # 1 is true, determine what - if any - action is required, and execute.
		"""
		# pass
		if self.handelTempChangeOnDevice and data.getTypeID() == ConfigConst.TEMP_SENSOR_TYPE:
			logging.info("Handle Temp change: %s - type ID: %s", str(self.handelTempChangeOnDevice), str(data.getTypeID()))

			ad = ActuatorData(typeID = ConfigConst.HVAC_ACTUATOR_TYPE)

			if data.getValue() > self.triggerHvacTempCeiling:
				ad.setCommand(ConfigConst.COMMAND_ON)
				ad.setValue(self.triggerHvacTempCeiling)

			elif data.getValue() < self.triggerHvacTempFloor:
				ad.setCommand(ConfigConst.COMMAND_ON)
				ad.setValue(self.triggerHvacTempFloor)

			else: 
				ad.setCommand(ConfigConst.COMMAND_OFF)

			self.handleActuatorCommandMessage(ad)

		
	def _handleUpstreamTransmission(self, resourceName: ResourceNameEnum, msg: str):
		"""
		Call this from handleActuatorCommandResponse(), handlesensorMessage(), and handleSystemPerformanceMessage()
		to determine if the message should be sent upstream. Steps to take:
		1) Check connection: Is there a client connection configured (and valid) to a remote MQTT or CoAP server?
		2) Act on msg: If # 1 is true, send message upstream using one (or both) client connections.
		"""
		logging.info("Upstream transmission invoked. Checking comm's integration.")
	
		# NOTE: If using MQTT, the following will attempt to publish the message to the broker
		if self.mqttClient:
			if self.mqttClient.publishMessage(resource = resourceName, msg = msg):
				logging.debug("Published incoming data to resource (MQTT): %s", str(resourceName))
			else:
				logging.warning("Failed to publish incoming data to resource (MQTT): %s", str(resourceName))
		
		# NOTE: If using CoAP, the following will attempt to PUT the message to the server
		if self.coapClient:
			if self.coapClient.sendPutRequest(resource = resourceName, payload = msg):
				logging.debug("Put incoming message data to resource (CoAP): %s", str(resourceName))
			else:
				logging.warning("Failed to put incoming message data to resource (CoAP): %s", str(resourceName))
		

File: Gateway_Device/java-components/src/main/java/programmingtheiot/gda/app/DeviceDataManager.java
/**
 * This class is part of the Programming the Internet of Things project.
 * 
 * It is provided as a simple shell to guide the student and assist with
 * implementation for the Programming the Internet of Things exercises,
 * and designed to be modified by the student as needed.
 */ 

package programmingtheiot.gda.app;

import java.time.OffsetDateTime;
import java.time.temporal.ChronoUnit;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.apache.commons.lang3.ObjectUtils.Null;

import programmingtheiot.common.ConfigConst;
import programmingtheiot.common.ConfigUtil;
import programmingtheiot.common.IActuatorDataListener;
import programmingtheiot.common.IDataMessageListener;
import programmingtheiot.common.ResourceNameEnum;

import programmingtheiot.data.ActuatorData;
import programmingtheiot.data.DataUtil;
import programmingtheiot.data.SensorData;
import programmingtheiot.data.BaseIotData;
import programmingtheiot.data.SystemPerformanceData;

import programmingtheiot.gda.connection.CloudClientConnector;
import programmingtheiot.gda.connection.CoapServerGateway;
import programmingtheiot.gda.connection.ICloudClient;
import programmingtheiot.gda.connection.IPersistenceClient;
import programmingtheiot.gda.connection.IPubSubClient;
import programmingtheiot.gda.connection.IRequestResponseClient;
import programmingtheiot.gda.connection.MqttClientConnector;
import programmingtheiot.gda.connection.RedisPersistenceAdapter;
import programmingtheiot.gda.connection.SmtpClientConnector;
import programmingtheiot.gda.system.SystemPerformanceManager;

/**
 * This class manages all the data and connections of GateWay Device App
 *
 */
public class DeviceDataManager implements IDataMessageListener
{
	// static
	
	private static final Logger _Logger = Logger.getLogger(DeviceDataManager.class.getName());
	
	// private var's
	
	private boolean enableMqttClient = true;
	private boolean enableCoapServer = false;
	private boolean enableCloudClient = true;
	private boolean enableSmtpClient = false;
	private boolean enablePersistenceClient = false;
	private boolean enableSystemPerf = false;
	
	private IActuatorDataListener actuatorDataListener = null;
	private IPubSubClient mqttClient = null;
	private ICloudClient cloudClient = null;
	private IPersistenceClient persistenceClient = null;
	private IRequestResponseClient smtpClient = null;
	private CoapServerGateway coapServer = null;
	private SystemPerformanceManager sysPerfMgr = null;
	private RedisPersistenceAdapter redisClientConnector = null;

	private ActuatorData latestHumidifierActuatorData = null;
	private ActuatorData latestLedData = null;
	private ActuatorData latestHumidifierActuatorResponse = null;

	private SensorData latestHumiditySensorData = null;
	private SensorData latestTemperatureSensorData = null;
	private SensorData latestPressureSensorData = null;

	private OffsetDateTime latestHumiditySensorTimeStamp = null;

	private boolean handleHumidityChangeOnDevice = false; // optional
	private int     lastKnownHumidifierCommand   = ConfigConst.OFF_COMMAND;

	
	private long    humidityMaxTimePastThreshold = 10; // seconds
	private float   nominalHumiditySetting   = 40.0f;
	private float   triggerHumidifierFloor   = 30.0f;
	private float   triggerHumidifierCeiling = 50.0f;
	
	// constructors
	
	public DeviceDataManager()
	{
		super();

		ConfigUtil configUtil = ConfigUtil.getInstance();

		// Getting all the configurations to able from Piotconfig.props
		this.enableMqttClient = configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, ConfigConst.ENABLE_MQTT_CLIENT_KEY);
		this.enableCoapServer = configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, ConfigConst.ENABLE_COAP_SERVER_KEY);
		this.enableCloudClient = configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, ConfigConst.ENABLE_CLOUD_CLIENT_KEY);
		this.enablePersistenceClient = configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, ConfigConst.ENABLE_PERSISTENCE_CLIENT_KEY);
		this.enableSystemPerf = configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, ConfigConst.ENABLE_SYSTEM_PERF_KEY);
		this.enablePersistenceClient = configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, ConfigConst.ENABLE_PERSISTENCE_CLIENT_KEY);

		// Getting paramaters to handle humidity values from Piotconfig.props
		this.handleHumidityChangeOnDevice =configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE, "handleHumidityChangeOnDevice");
		this.humidityMaxTimePastThreshold =configUtil.getInteger(ConfigConst.GATEWAY_DEVICE, "humidityMaxTimePastThreshold");
		this.nominalHumiditySetting =configUtil.getFloat(ConfigConst.GATEWAY_DEVICE, "nominalHumiditySetting");
		this.triggerHumidifierFloor =configUtil.getFloat(ConfigConst.GATEWAY_DEVICE, "triggerHumidifierFloor");
		this.triggerHumidifierCeiling =configUtil.getFloat(ConfigConst.GATEWAY_DEVICE, "triggerHumidifierCeiling");

		if (this.humidityMaxTimePastThreshold < 10 || this.humidityMaxTimePastThreshold > 7200) {
			this.humidityMaxTimePastThreshold = 30;
		}

		initManager();
		
		initConnections();
	}
	
	public DeviceDataManager(
		boolean enableMqttClient,
		boolean enableCoapClient,
		boolean enableCloudClient,
		boolean enableSmtpClient,
		boolean enablePersistenceClient)
	{
		super();
		
		initConnections();
	}
	

	// public methods
	
	@Override
	public boolean handleActuatorCommandResponse(ResourceNameEnum resourceName, ActuatorData data)
	{
		if (data != null)
		{
			_Logger.info("Handling actuator response: "+data.getName()+" "+data.getValue());

			if(data.hasError()){
				_Logger.warning("Error flag set for the Actuation data instance");
			}
			return true;
		}
		else{
			return false;
		}
	} 

	@Override
	public boolean handleActuatorCommandRequest(ResourceNameEnum resourceName, ActuatorData data)
	{
		if(data != null)
		{
			_Logger.info("Handling Actuator request received" + data.getCommand());
			_Logger.log(Level.FINE, "Actuator request received: {0}.Message: {1}",
			new Object[]{resourceName.getResourceName(), Integer.valueOf((data.getCommand()))});

			if(data.hasError()){
				_Logger.warning("Error flag is set for Actuator Data Instance");
			}
			int qos = ConfigConst.DEFAULT_QOS;

			this.sendActuatorCommandtoCda(resourceName, data);
			return true;
		}
		else{
			return false;
		}
	}

	@Override
	public boolean handleIncomingMessage(ResourceNameEnum resourceName, String msg)
	{
		if(resourceName != null && msg != null)
		{
			try {
				if(resourceName == ResourceNameEnum.CDA_ACTUATOR_CMD_RESOURCE){
					_Logger.info("Handiling incomming general Message: " + msg);

					ActuatorData ad = DataUtil.getInstance().jsonToActuatorData(msg);
					String jsonData = DataUtil.getInstance().actuatorDataToJson(ad);
					
						if(this.mqttClient != null){
							_Logger.info("Publishing data to MQTT broker: " + jsonData);
							return this.mqttClient.publishMessage(resourceName, jsonData, 0);
						}

					}
					else{
						_Logger.warning("Failed to parse incomming message. Unknown type: " + msg);
						return false;
					}

				}catch(Exception e){
					_Logger.log(Level.WARNING, "Failed to process incomming message for resource: " + resourceName, e);
				}
		}
		else{
			_Logger.warning("Incomming message has no data. Ignoring for resource: " + resourceName);
		}
		return false;
	}

	@Override
	public boolean handleSensorMessage(ResourceNameEnum resourceName, SensorData data)
	{
		
		if(data != null)
		{
			_Logger.info("Handling sensor message: " + data.getName());

			if(data.hasError())
			{
				_Logger.warning("Error flag set for the SensorData instance");
			}

			String jsonData = DataUtil.getInstance().sensorDataToJson(data);
			_Logger.info("JSON [SensorData] -> " + jsonData);

			int qos = ConfigConst.DEFAULT_QOS;

			if(this.enablePersistenceClient && this.persistenceClient != null)
			{
				this.persistenceClient.storeData(resourceName.getResourceName(), qos, data);
			}

			this.handleIncomingDataAnalysis(resourceName, data);
			this.handleUpstreamTransmission(resourceName, jsonData, qos);

			return true;
		}
		else{
			return false;
		}
	}

	// Method to handle system performance data
	@Override
	public boolean handleSystemPerformanceMessage(ResourceNameEnum resourceName, SystemPerformanceData data)
	{
		if(data != null)
		{
			_Logger.info("Handling System Performance message: " + data.getName());

			if (data.hasError())
			{
				_Logger.warning("Error flag set for SystemPerformanceData instance");
			}

			String jsonData = DataUtil.getInstance().systemPerformanceDataToJson(data);
			
			int qos = ConfigConst.DEFAULT_QOS;

			this.handleUpstreamTransmission(resourceName, jsonData, qos);

			return true;
		}
		else{
			return false;
		}
	}
	
	public void setActuatorDataListener(String name, IActuatorDataListener listener)
	{
		if(listener != null){
			this.actuatorDataListener = listener;
		}
	}
	
	public void startManager()
	{
		_Logger.info("*******Device Data Manager Started Successfully*******");

		// Adding MQTT client and subscribing to a topic
		if (this.mqttClient != null) 
		{
			if (this.mqttClient.connectClient()) 
			{
				_Logger.info("Successfully connected MQTT client to broker.");
			}
			else 
			{
				_Logger.severe("Failed to connect MQTT client to broker.");
			}
		}

		if(this.cloudClient != null)
		{
			if(this.cloudClient.connectClient())
			{
				_Logger.info("Successfully connected to Cloud Service Provider");
			}
			else{
				_Logger.severe("Failed to connect to Cloud Service Provider");
			}
		}

		if(this.redisClientConnector != null)
		{
			if(this.redisClientConnector.connectClient())
			{
				_Logger.info("Successfully connected to Redis Data Base");
			}
			else{
				_Logger.severe("Failed to connet to Redis Data Base");
			}
		}
		if(this.enableCoapServer && this.coapServer != null){
			if(this.coapServer.startServer()){
				_Logger.info("CoAP server started.");
			}else{
				_Logger.severe("Failed to start CoAP server. Check log file for details.");
			}
		}

		if(this.sysPerfMgr != null)
		{
			_Logger.info("*******System Performance Manager Started Successfully*******");
			this.sysPerfMgr.startManager();
		}
	}

	
	public void stopManager()
	{
		if (this.mqttClient != null)
		{
			// Unsubscribing and disconnecting MQTT client
			this.mqttClient.unsubscribeFromTopic(ResourceNameEnum.GDA_MGMT_STATUS_MSG_RESOURCE);
			this.mqttClient.unsubscribeFromTopic(ResourceNameEnum.CDA_ACTUATOR_RESPONSE_RESOURCE);
			this.mqttClient.unsubscribeFromTopic(ResourceNameEnum.CDA_SENSOR_MSG_RESOURCE);
			this.mqttClient.unsubscribeFromTopic(ResourceNameEnum.CDA_SYSTEM_PERF_MSG_RESOURCE);

			this.mqttClient.disconnectClient();
			
			if (this.mqttClient.disconnectClient())
			{
				_Logger.info("Successfully disconnected MQTT client from broker.");
			} 
			else
			{
				_Logger.severe("Failed to disconnect MQTT client from broker.");	
			}
		}

		if (this.cloudClient != null)
		{
			this.cloudClient.unsubscribeFromCloudEvents(ResourceNameEnum.CDA_ACTUATOR_CMD_RESOURCE);
			this.cloudClient.disconnectClient();
		}
		if(this.cloudClient.disconnectClient()){
			_Logger.info("Success disconnected from Cloud service provider");
		}else{
			_Logger.severe("Failed to disconnect from Cloud Service Provider");
		}

		if(this.enableCoapServer && this.coapServer != null){
			if(this.coapServer.stopServer()){
				_Logger.info("CoAP server stopped.");
			}else{
				_Logger.severe("Failed to stop CoAP server. Check log file for details.");
			}
		}

		if(this.sysPerfMgr != null)
		{
			this.sysPerfMgr.stopManager();
			_Logger.info("******Syatem performance Manager Stopped Successfully******");
		}
	}

	
	// private methods
	
	/**
	 * Initializes the enabled connections. This will NOT start them, but only create the
	 * instances that will be used in the {@link #startManager() and #stopManager()) methods.
	 * 
	 */

	private void initConnections()
	{
	}


	private void initManager()
	{
		ConfigUtil configUtil = ConfigUtil.getInstance();

		this.enableSystemPerf = configUtil.getBoolean(ConfigConst.GATEWAY_DEVICE , ConfigConst.ENABLE_SYSTEM_PERF_KEY);

		if(this.enableSystemPerf){
			this.sysPerfMgr = new SystemPerformanceManager();
			this.sysPerfMgr.setDataMessageListener(this);
		}

		if(this.enableMqttClient){
			this.mqttClient = new MqttClientConnector();
			this.mqttClient.setDataMessageListener(this);
		}

		if(this.enableCoapServer){
			this.coapServer = new CoapServerGateway(this);
		}

		if(this.enableCloudClient){
			this.cloudClient = new CloudClientConnector();
			this.cloudClient.setDataMessageListener(this);
		}

		if(this.enablePersistenceClient){
			this.redisClientConnector = new RedisPersistenceAdapter();
			// this.redisClientConnector.setPersistanceListener(this);
		}
	}
	

	private void handleIncomingDataAnalysis(ResourceNameEnum resourceName, ActuatorData data)
	{
		_Logger.info("Analysing incoming actuator data: " + data.getName());

		if(data.isResponseFlagEnabled()){

		}else{
			if(this.actuatorDataListener != null)
			{
				this.actuatorDataListener.onActuatorDataUpdate(data);
			}
		}
	}

	private boolean handleIncomingDataAnalysis(ResourceNameEnum resourceName, SensorData data)
	{
		if(data != null)
		{
			_Logger.info("Analysing incoming SensorData");
			if (data.getTypeID() == ConfigConst.HUMIDITY_SENSOR_TYPE)
			{
				handleHumiditySensorAnalysis(resourceName, data);
			}
			return true;
		}
		else{
			return false;
		}
	}

	private void handleHumiditySensorAnalysis(ResourceNameEnum resource, SensorData data)
	{
		//
		// NOTE: INCOMPLETE and VERY BASIC CODE SAMPLE. Not intended to provide a solution.
		//
		_Logger.fine("Analyzing humidity data from CDA: " + data.getLocationID() + ". Value: " + data.getValue());
	
		boolean isLow  = data.getValue() < this.triggerHumidifierFloor;
		boolean isHigh = data.getValue() > this.triggerHumidifierCeiling;
	
		if (isLow || isHigh) {
			_Logger.fine("Humidity data from CDA exceeds nominal range.");
		
			if (this.latestHumiditySensorData == null) {
				// set properties then exit - nothing more to do until the next sample
				this.latestHumiditySensorData = data;
				this.latestHumiditySensorTimeStamp = getDateTimeFromData(data);
			
				_Logger.info("Starting humidity nominal exception timer. Waiting for seconds: " + this.humidityMaxTimePastThreshold);
			
				return;
			} else {
				OffsetDateTime curHumiditySensorTimeStamp = getDateTimeFromData(data);
			
				long diffSeconds = ChronoUnit.SECONDS.between(this.latestHumiditySensorTimeStamp, curHumiditySensorTimeStamp);
			
				_Logger.info("Checking Humidity value exception time delta: " + diffSeconds);
			
				if (diffSeconds >= this.humidityMaxTimePastThreshold) 
				{
					ActuatorData ad = new ActuatorData();
					ad.setName(ConfigConst.HUMIDIFIER_ACTUATOR_NAME);
					ad.setLocationID(data.getLocationID());
					ad.setTypeID(ConfigConst.HUMIDIFIER_ACTUATOR_TYPE);
					ad.setValue(this.nominalHumiditySetting);
				
					if (isLow) 
					{
						ad.setCommand(ConfigConst.ON_COMMAND);
					}
					else if (isHigh)
					{
						ad.setCommand(ConfigConst.OFF_COMMAND);
					}
				
					_Logger.info("Humidity exceptional value reached. Sending actuation event to CDA: " + ad);
				
					this.lastKnownHumidifierCommand = ad.getCommand();
					sendActuatorCommandtoCda(ResourceNameEnum.CDA_ACTUATOR_CMD_RESOURCE, ad);
				
					// set ActuatorData and reset SensorData (and timestamp)
					this.latestHumidifierActuatorData = ad;
					this.latestHumiditySensorData = null;
					this.latestHumiditySensorTimeStamp = null;
				}
			}
		} else if (this.lastKnownHumidifierCommand == ConfigConst.ON_COMMAND) {
			// check if we need to turn off the humidifier
			if (this.latestHumidifierActuatorData != null) {
				// check the value - if the humidifier is on, but not yet at nominal, keep it on
				if (this.latestHumidifierActuatorData.getValue() >= this.nominalHumiditySetting) {
					this.latestHumidifierActuatorData.setCommand(ConfigConst.OFF_COMMAND);
				
					_Logger.info("Humidity nominal value reached. Sending OFF actuation event to CDA: " + this.latestHumidifierActuatorData);
				
					sendActuatorCommandtoCda(ResourceNameEnum.CDA_ACTUATOR_CMD_RESOURCE, this.latestHumidifierActuatorData);
				
					// reset ActuatorData and SensorData (and timestamp)
					this.lastKnownHumidifierCommand = this.latestHumidifierActuatorData.getCommand();
					this.latestHumidifierActuatorData = null;
					this.latestHumiditySensorData = null;
					this.latestHumiditySensorTimeStamp = null;
				} else {
					_Logger.info("Humidifier is still on. Not yet at nominal levels (OK).");
				}
			} else {
				// shouldn't happen, unless some other logic
				// nullifies the class-scoped ActuatorData instance
				_Logger.warning("ERROR: ActuatorData for humidifier is null (shouldn't be). Can't send command.");
			}
		}
	}

	private void sendActuatorCommandtoCda(ResourceNameEnum resource, ActuatorData data)
	{
		// NOTE: This is how an ActuatorData command will get passed to the CDA
		// when the GDA is providing the CoAP server and hosting the appropriate
		// ActuatorData resource. It will typically be used when the OBSERVE
		// client (the CDA, assuming the GDA is the server and CDA is the client)
		// has sent an OBSERVE GET request to the ActuatorData resource.

		if (this.actuatorDataListener != null) {
			this.actuatorDataListener.onActuatorDataUpdate(data);
		}
		
		// NOTE: This is how an ActuatorData command will get passed to the CDA
		// when using MQTT to communicate between the GDA and CDA
		if (this.enableMqttClient && this.mqttClient != null) {
			String jsonData = DataUtil.getInstance().actuatorDataToJson(data);
			
			if (this.mqttClient.publishMessage(resource, jsonData, ConfigConst.DEFAULT_QOS)) {
				_Logger.info("Published ActuatorData command from GDA to CDA: " + data.getCommand());
			} else {
				_Logger.warning("Failed to publish ActuatorData command from GDA to CDA: " + data.getCommand());
			}
		}
	}

	private OffsetDateTime getDateTimeFromData(BaseIotData data)
	{
		OffsetDateTime odt = null;
	
		try {
			odt = OffsetDateTime.parse(data.getTimeStamp());
		} catch (Exception e) {
			_Logger.warning("Failed to extract ISO 8601 timestamp from IoT data. Using local current time.");
		
			// TODO: this won't be accurate, but should be reasonably close, as the CDA will
			// most likely have recently sent the data to the GDA
			odt = OffsetDateTime.now();
		}
	
		return odt;
	}

	private boolean handleIncomingDataAnalysis(ResourceNameEnum resourceName, SystemPerformanceData data)
	{
		if(data != null)
		{
			_Logger.info("Analysing incoming SystemPerformanceData");

			return true;
		}
		else{
			return false;
		}
	}

	// Method to send data to cloud
	private boolean handleUpstreamTransmission(ResourceNameEnum resourceName, String jsonData, int qos)
	{
		
		if (jsonData != null)
		{
			_Logger.info("Sending JSON Data to cloud service: " + resourceName);
			
			if(this.cloudClient != null)
			{
				if(resourceName == ResourceNameEnum.CDA_SENSOR_MSG_RESOURCE)
				{
					SensorData data = DataUtil.getInstance().jsonToSensorData(jsonData);
					this.cloudClient.sendEdgeDataToCloud(resourceName, data);
					return true;
				}
				else if (resourceName == ResourceNameEnum.GDA_SYSTEM_PERF_MSG_RESOURCE)
				{
					SystemPerformanceData data = DataUtil.getInstance().jsonToSystemPerformanceData(jsonData);
					this.cloudClient.sendEdgeDataToCloud(resourceName, data);
					return true;
				}
				else if (resourceName == ResourceNameEnum.CDA_SYSTEM_PERF_MSG_RESOURCE)
				{
					SystemPerformanceData data = DataUtil.getInstance().jsonToSystemPerformanceData(jsonData);
					this.cloudClient.sendEdgeDataToCloud(resourceName, data);
					return true;
				}
				else
				{
					_Logger.warning("Resource Match Not Found " + resourceName);
					return false;
				}
			}
			else{
				_Logger.info("Cloud client no connected to broker: ");
				return false;
			}
		}
		else{
			_Logger.warning("No JSON data received. Failed to up stream data: " + resourceName);
			return false;
		}
	}

}
